# GenoPipe: identifying the genotype of origin within (epi)genomic datasets

## Olivia W Lang, Divyanshi Srivastava, B Franklin Pugh, William K M Lai

*Department of Molecular Biology and Genetics, Cornell University, Ithaca, NY 14853, USA.*

*Department of Biochemistry & Molecular Biology, Pennsylvania State University, University Park, PA, 16801, USA.*

*Department of Computational Biology, Cornell University, Ithaca, NY 14850, USA.*

*Cornell Institute of Biotechnology, Cornell University, Ithaca, NY 14850, USA.*

### [:house: GenoPipe Website Homepage :house:](https://www.genopipe.org)
### PMID : [37933851](https://pubmed.ncbi.nlm.nih.gov/37933851/)

## Overview
Confidence in experimental results is critical for discovery. As the scale of data generation in genomics has grown exponentially, experimental error has likely kept pace despite the best efforts of many laboratories. Technical mistakes can and do occur at nearly every stage of a genomics assay (i.e., cell line contamination, reagent swapping, tube mislabelling, etc.) and are often difficult to identify post-execution. However, the DNA sequenced in genomic experiments contains certain markers (e.g., indels) encoded within and can often be ascertained forensically from experimental datasets. We developed the Genotype validation Pipeline (GenoPipe), a suite of heuristic tools that operate together directly on raw and aligned sequencing data from individual high-throughput sequencing experiments to characterize the underlying genome of the source material. We demonstrate how GenoPipe validates and rescues erroneously annotated experiments by identifying unique markers inherent to an organismâ€™s genome (i.e., epitope insertions, gene deletions, and SNPs).


